MPAS dycore + Atmosphere Generalization & Update RAP tests to FHMAX=12

dustinswales opened the pull request:
<!-- INSTRUCTIONS: 
- PLEASE READ/FOLLOW THE DIRECTIONS IN EACH SECTION
- Complete the 'Commit Queue Requirements' below
- Please use github markup as much as possible (https://docs.github.com/en/get-started/writing-on-github)
- Please leave your PR in a draft state until all underlying work is completed.
-->
## Commit Queue Requirements:
<!--
- Please complete the items that follow this.
- Please "check off" completed items. Use [X] for a filled in checkbox or leave it [ ] for an empty checkbox
- Your PR will not be considered until all requirements are met.
- THIS IS YOUR RESPONSIBILITY
 -->
- [x] Fill out all sections of this template.
- [x] All sub component pull requests have been reviewed by their code managers.
- [x] Run the full Intel+GNU RT suite (compared to current baselines) on either Hera/Derecho/Hercules
- [x] Commit 'test_changes.list' from previous step
---
## Description:
This pull request introduces the dynamical core from the Model for Prediction Across Scales (MPAS) and contains changes to generalize atmospheric component of the UFS Weather Model to accomodate multiple dynamical cores. 

This PR supersedes #2451 

PR #2634 has also been combined into this PR. It updates the default value of FHMAX to 12 (rather than 24) in export_rap. It also updates the expected LIST_FILES, RESTART_INTERVAL, and OUTPUT_FH in several following RAP RTs to reflect the 12-hour FHMAX.

- New baseline case control_gfs_mpas is currently set to check MPAS log output: LIST_FILES="mpas_log.txt"

### Commit Message:
```
* UFSWM - Initial implementation of the MPAS dynamical core in the UWM atmosphere; shorten RAP tests to 12 hrs
  * CMEPS - Changes for the renaming of FV3 to UFSATM 
  * noahmp - Changes for the renaming of FV3 to UFSATM 
  * UFSATM - Initial implementation of the MPAS dynamical core in the atmospheric component of the UWM.
```

### Priority:
* Normal

## Git Tracking
### UFSWM:
* Closes #2517 
* Closes #2611 
* Closes #2612 
* Closes #1290 (RAP tests)
* Partially address #2524 
* Partially address #2523 
* Partially address #2522 

### Sub component Pull Requests:
* FV3: NOAA-EMC/fv3atm#982
* CMEPS: NOAA-EMC/CMEPS#145
* NoahMP: NOAA-EMC/noahmp#14
* ccpp/physics: ufs-community/ccpp-physics#301

### UFSWM Blocking Dependencies:
* None

### Documentation:
* No documentation update is required for this PR. These are many more changes needed to create MPAS enabled forecasts, at which point the documentation will need to be updated.

---
## Changes
### Regression Test Changes (Please commit test_changes.list):
* PR Adds New Tests/Baselines (_control_rrfs_mpas_ and _control_gfs_mpas_) *Updated on 08/28/2025 *
* No Baseline Changes to existing tests

### Input data Changes:
* New input data.
On Hera:
- /scratch4/NAGAPE/epic/role-epic/UFS-WM_RT/NEMSfv3gfs/input-data-20250507/MPAS/INPUT/120km/CONUS/
- /scratch4/NAGAPE/epic/role-epic/UFS-WM_RT/NEMSfv3gfs/input-data-20250507/MPAS/INPUT/120km/Global/

### Library Changes/Upgrades:
N/A

<!-- STOP!!! THE FOLLOWING IS FOR CODE MANAGERS ONLY. PLEASE DO NOT FILL OUT -->
## Testing Log:
- RDHPCS
  - [x] Hera
  - [x] Orion
  - [x] Hercules
  - [x] GaeaC6
  - [x] Derecho
  - [x] Ursa
- WCOSS2
  - [x] Dogwood/Cactus
  - [x] Acorn
- [x] CI
- [x] opnReqTest (complete task if unnecessary)

danrosen25 commented on CMakeLists.txt:
```diff
@@ -202,6 +210,9 @@ endif()
   add_subdirectory(FV3)
 endif()
+if (MPAS)
+  add_subdirectory(FV3)
```
I recommend renaming the repository and submodule folder. When renaming a repository via GitHub the platform will create a redirect link.

dustinswales replied:
Will do.

climbfuji replied:
Agreed, this is very confusing: if MPAS --> add_subdirectory FV3

dustinswales replied:
Done.

danrosen25 commented on cmake/configure_apps.cmake:
```diff
@@ -13,7 +13,7 @@
 ### Configure Application Components
 ###############################################################################
-if(APP MATCHES "^(ATM|ATMW|ATMWM|ATMAQ|ATML|ATMF|ATML-LM4)$")
+if(APP MATCHES "^(ATM|ATMW|ATMWM|ATMAQ|ATML|ATMF|ATML-LM4|ATMMPAS)$")
```
At some point this should be rewritten because of increasing complexity.

dustinswales replied:
Agreed.

danrosen25 commented on tests/fv3_conf/rrfs_mpas_run.IN:
```diff
@@ -0,0 +1,58 @@
+mkdir INPUT RESTART
+
+# MPAS Dycore
+cp /scratch1/BMC/gmtb/Dustin.Swales/MPAS/INPUT/@[MPAS_RESOLUTION]km/CONUS/* .
```
This assumes that all MPAS runs using rrfs_mpas_run will have input in a folder labeled CONUS

dustinswales replied:
Correct. I added that this PR has new input data in the description.
Once the data is staged, these will be replaced with the data in their correct location.

danrosen25 commented on tests/rt.conf:
This file is missing a newline at the end of the file. This isn't usually a problem but it can because Unix utilities may utilize the newline character.

danrosen25 commented on tests/tests/control_gfs_mpas:
Also needs a newline at the end of file.

dustinswales replied:
Done.

danrosen25 commented on tests/tests/control_rrfs_mpas:
Also needs a newline at the end of the file.

dustinswales replied:
Done.

danrosen25 commented on tests/tests/control_rrfs_mpas_debug:
Also needs a newline at the end of the file.

dustinswales replied:
Done.

danrosen25 reviewed (commented):
The restructuring FV3 to UFSATM work looks good. I think it might be worth discussing the growth of ufs-weather-model applications and if there is a better process for this going forward. Some of the files are missing newline terminators.

DusanJovic-NOAA commented on CMakeLists.txt:
```diff
@@ -180,6 +182,12 @@ find_package(bacio 2.4.0 REQUIRED)
+if(MPAS)
+  find_package(PnetCDF REQUIRED COMPONENTS Fortran)
+  find_package(PIO REQUIRED COMPONENTS Fortran C)
+  set(FV3 OFF)
```
Isn't FV3 OFF by default, why is this needed?

dustinswales replied:
You are correct. Removed.

DusanJovic-NOAA commented on CMakeLists.txt:
```diff
@@ -44,6 +45,7 @@ include(cmake/configure_apps.cmake)
 message("")
 message("FMS .............. ${FMS}")
 message("FV3_dy ........... ${FV3}")
+message("MPAS_dy .......... ${MPAS}")
```
Consider removing '_dy' suffix here. Other components do not have any suffix.

dustinswales replied:
Done.

DusanJovic-NOAA commented on tests/tests/control_gfs_mpas:
```diff
@@ -0,0 +1,19 @@
+
+export CNTL_DIR=control_gfs_mpas
+
+export LIST_FILES=""
```
Why is 'LIST_FILES' empty? Are MPAS tests not reproducible?

dustinswales replied:
There isn't any output yet.
I will add the MPAS logfile here. At least that will alert us if anyone broke what is working up until this point. @DusanJovic-NOAA What do you think?

DusanJovic-NOAA replied:
I'm not sure what kind of log files MPAS creates, but the files listed here are saved as baselines, so they must be reproducible from run to run. If log files contain any timing info or anything that might differ between runs, then that will not work.

dustinswales replied:
@DusanJovic-NOAA I added a new log file for MPAS. It contains all of the logging information from the MPAS dycore initialization through the first call to the dycore (reproducible). 
In the future we will have actual output files to compare, and this could go away.

uturuncoglu commented on tests/default_vars.sh:
```diff
@@ -1799,6 +2125,7 @@ export_hafs ()
 {
 export_fv3_v16
 export FV3=true
+export MPAS=false
```
If FV3 OFF by default in the top level cmake file. Why it is true by default in here. So, it would be nice to make them consistent.

dustinswales replied:
Good point. Removed.
This default_vars.sh file is a total mess and needs some love, albeit out-of-scope with this PR.

uturuncoglu reviewed (commented):
I'll take a look to FV3 changes separately. BTW, maybe it would be nice to rename the component as UFSATM at this point.

climbfuji commented on tests/rt.conf:
```diff
@@ -503,5 +503,13 @@ COMPILE | s2sw_pdlib_debug | gnu | -DAPP=S2SW -DCCPP_SUITES=FV3_GFS_v17_coupled_
 ### CDEPS Data Atmosphere test ###
+COMPILE | datm_cdeps | gnu | -DAPP=NG-GODAS | + hera hercules | fv3 |
+RUN | datm_cdeps_control_cfsr                           | + hera hercules                      | baseline |
+
```
Is this related to the MPAS addition and generalization?

dustinswales replied:
Sorry, this was a merge conflict that I didn't resolve correctly.
Fixed now.

climbfuji commented on tests/tests/control_gfs_mpas:
```diff
@@ -0,0 +1,19 @@
+export CCPP_SUITE=MPAS_RRFS
+export INPUT_NML=control_gfs_mpas.nml.IN
+export IC_FILENAME=x1.40962.init.nc
+export GRID_FILENAME=x1.40962.graph.info.part.
```
```suggestion
export GRID_FILENAME=x1.40962.graph.info.part.

```

climbfuji commented on tests/tests/control_rrfs_mpas:
```diff
@@ -0,0 +1,21 @@
+export INPUT_NML=control_rrfs_mpas.nml.IN
+export IC_FILENAME=x1.40962.init.nc
+export LBC_FILENAME=x1.40962.lbc.nc
+export GRID_FILENAME=conus.120km.graph.info.part.
```
```suggestion
export GRID_FILENAME=conus.120km.graph.info.part.

```

climbfuji commented on tests/tests/control_rrfs_mpas_debug:
```diff
@@ -0,0 +1,21 @@
+export INPUT_NML=control_rrfs_mpas.nml.IN
+export IC_FILENAME=x1.40962.init.nc
+export LBC_FILENAME=x1.40962.lbc.nc
+export GRID_FILENAME=conus.120km.graph.info.part.
```
```suggestion
export GRID_FILENAME=conus.120km.graph.info.part.

```

BrianCurtis-NOAA commented:
Is there anywhere thats blocking UFSATM and MPAS being selected/enabled at the same time?

dustinswales commented:
> Is there anywhere thats blocking UFSATM and MPAS being selected/enabled at the same time?

@BrianCurtis-NOAA If FV3 or MPAS dycore, then UFSATM is enabled. The MPAS and FV3 macros are used within UFSATM to control which dycore is selected.
I just added a  check to ensure the multiple dycores are not being sleected.

dustinswales commented:
@danrosen25 @uturuncoglu @DomHeinzeller Thanks for the comments. I made the name change from FV3 -> UFSATM. This was straightforward, but required some small changes in CMEPS (@DeniseWorthen) and noahmp (@barlage).

dustinswales commented:
@DomHeinzeller @danrosen25 @DusanJovic-NOAA @BrianCurtis-NOAA 
Sorry for the delay, but I ran into some issues updating and fully migrating the name change of FV3->UFSATM.
All tests pass, except new MPAS test, which is expected. See RT log file (hera) and test_changes.list.

- MPAS tests are now using GNU, instead of Intel. Not sure what's going on with the stack, but now when I try to build the MPAS RT with the latest Intel on spack-stackv1.9.1, I get a buildtime error in the MPAS framework, ~"icx cannot use -threads -cxxlib". I couldn't find the source of them after a day or so of sleuthing, so I moved to GNU for the time being.

- FV3 -> UFSATM. New PRs(3): Noahmp, CMEPS, and ccpp-physics. See PR in these repos for more detail.

- Address reviewers comments. @climbfuji The only thing I didn't do was unify MPAS_init and GFS_init. I updated the description to "partially address" #2524, instead of closing it. It's on my list to clean this up with the next round of changes.

jkbk2004 commented:
@dustinswales scratch1/2 are not available anymore. can you prepare the input files somewhere on scratch3/4?

jkbk2004 commented:
@dustinswales can you sync up branch? we can start working on this pr.

jkbk2004 commented:
@BrianCurtis-NOAA new input files are available on /scratch4/NAGAPE/epic/role-epic/UFS-WM_RT/NEMSfv3gfs/input-data-20250507/MPAS. @dustinswales @DusanJovic-NOAA @junwang-noaa we can still move on with this pr for testing. An outstanding question is about where to host UFSATM repository. EMC repository should be ok until MPAS component features and develop branch fully get matured. But should we consider to migrate to ufs-community as a general naming UFSATM is adopted?

dustinswales commented:
> @BrianCurtis-NOAA new input files are available on /scratch4/NAGAPE/epic/role-epic/UFS-WM_RT/NEMSfv3gfs/input-data-20250507/MPAS. @dustinswales @DusanJovic-NOAA @junwang-noaa we can still move on with this pr for testing. An outstanding question is about where to host UFSATM repository. EMC repository should be ok until MPAS component features and develop branch fully get matured. But should we consider to migrate to ufs-community as a general naming UFSATM is adopted?

@jkbk2004 Good questions. 
In the past we've (@junwang-noaa @DusanJovic-NOAA and others) discussed renaming the repository from NOAA-EMC/fv3atm to NOAA-EMC/ufsatm. However, we didn't discuss moving the repository out of the NOAA-EMC organization. (I don't see a reason to move the repo, but other may want to chime)

jkbk2004 commented:
@gspetro-NOAA we will combine in #2634 to this pr.

gspetro-NOAA commented:
Updated baselines for the changed RAP tests from PR #2634 have been created at the following locations: 
-  Hercules: /work2/noaa/epic/gpetro/hercules/RTs/ufs-wm/stmp/gpetro/FV3_RT/REGRESSION_TEST
- Orion: /work2/noaa/epic/gpetro/orion/RTs/ufs-wm/stmp/gpetro/FV3_RT/REGRESSION_TEST
- Ursa: /scratch4/NAGAPE/epic/Gillian.Petro/stmp/RT_BASELINE/Gillian.Petro/FV3_RT/REGRESSION_TEST
- Hera: /scratch3/NCEPDEV/stmp/Gillian.Petro/RT_BASELINE/Gillian.Petro/FV3_RT
- GaeaC6: /gpfs/f6/bil-fire8/world-shared/Gillian.Petro/RT_BASELINE/Gillian.Petro/FV3_RT/REGRESSION_TEST
- Derecho: /glade/derecho/scratch/gpetro/FV3_RT/REGRESSION_TEST

jkbk2004 commented:
@dustinswales issue is with a direct git clone option in cmake file. @junwang-noaa @DusanJovic-NOAA @BrianCurtis-NOAA MPAS core has a different approach to pull source code.
```
CMake Error at UFSATM/mpas/MPAS-Model/src/core_atmosphere/CMakeLists.txt:109 (message):
  Git clone failed with error: Cloning into
  '/work/noaa/epic/jongkim/UFS-RT/hercules/rt-2780/UFSATM/mpas/MPAS-Model/src/core_atmosphere/physics/physics_mmm'...


  fatal: unable to access 'https://github.com/NCAR/MMM-physics/': Failed to
  connect to github.com port 443 after 131046 ms: Could not connect to server
```
Also, an issue with column 6 at https://github.com/dustinswales/ufs-weather-model/blob/feature/mpas_nuopc_init/tests/rt.conf#L505. It conflicts with https://github.com/dustinswales/ufs-weather-model/blob/feature/mpas_nuopc_init/tests/rt.sh#L1233. Too many "fv3" footprints at tests/rt.sh level. So I recommend to keep using "fv3" at the rt.conf column 6 until we completely generalize test script with ufsatm option. @FernandoAndrade-NOAA @gspetro-NOAA FYI

dustinswales commented:
@jkbk2004 I fixed the git clone issue and changed "mpas" to "fv3" in rt.conf (There will be more PRs that I could address this in)

NickSzapiro-NOAA commented on CMakeLists.txt:
```diff
@@ -164,7 +166,7 @@ if(FMS)
     add_library(fms ALIAS FMS::fms_r8)
   endif()
-  if(APP MATCHES "^(ATM|ATMAERO|ATMAQ|ATMWM|ATMW|ATML|ATMF|ATML-LM4|LND-LM4|HAFS|HAFS-ALL)$")
+  if(APP MATCHES "^(ATM|ATMAERO|ATMAQ|ATMWM|ATMW|ATML|ATMF|ATMMPAS|ATML-LM4|LND-LM4|HAFS|HAFS-ALL)$")
```
Why does MPAS need FMS?

dustinswales replied:
UFSATM/ccpp/data/GFS_typedefs.F90, which is used by both FV3 and MPAS, requires FMS.

NickSzapiro-NOAA commented on cmake/configure_apps.cmake:
```diff
@@ -31,6 +31,14 @@ if(APP MATCHES "^(ATM|ATMW|ATMWM|ATMAQ|ATML|ATMF|ATML-LM4)$")
+    set(MPAS     ON  CACHE BOOL "Enable MPAS dycore"         FORCE)
+    # DJS2025: FV3=ON has been factored out of all ATM applications, this line
+    #          reverts that when MPAS dycore is selected. Going forward, as diff. applications
+    #          rely on different dycores, this logic will need to expand to choose the
+    #          correct dycore.
+    set(FV3      OFF CACHE BOOL "Disable FV3 dycore"         FORCE)
```
set FMS off too, no?

NickSzapiro-NOAA commented:
@dustinswales I don't know where to add review comment but there's a leftover from git merge in:
https://github.com/ufs-community/MPAS-Model/blob/af8d9e86356b354059a5b9ec461360c8b3c1e9b2/src/core_atmosphere/Registry.xml#L1303

dustinswales commented:
> @dustinswales I don't know where to add review comment but there's a leftover from git merge in: https://github.com/ufs-community/MPAS-Model/blob/af8d9e86356b354059a5b9ec461360c8b3c1e9b2/src/core_atmosphere/Registry.xml#L1303

@NickSzapiro-NOAA Good catch. Removed.

DusanJovic-NOAA commented:
Does this PR require new baselines?

jkbk2004 commented:
> Does this PR require new baselines?

Not experiment output baseline. It sets up MPAS log as baseline: mpas_log.txt. I think we expect more PRs. @dustinswales any comment?

dustinswales commented:
Correct. Just the MPAS log for now. There will be subsequent PRs as we work through the implementation.
The PR description states "adds new Tests". I will adjust to say "adds Tests/Baselines".

DusanJovic-NOAA commented:
`test_changes.list' shows that these tests need new baselines:
```
control_gfs_mpas gnu
rap_control intel
rap_decomp intel
rap_2threads intel
rap_restart intel
rap_sfcdiff intel
rap_sfcdiff_decomp intel
rap_sfcdiff_restart intel
rap_control_dyn32_phy32 intel
rap_2threads_dyn32_phy32 intel
rap_restart_dyn32_phy32 intel
rap_control gnu
rap_decomp gnu
rap_2threads gnu
rap_restart gnu
rap_sfcdiff gnu
rap_sfcdiff_decomp gnu
rap_sfcdiff_restart gnu
rap_control_dyn32_phy32 gnu
rap_2threads_dyn32_phy32 gnu
rap_restart_dyn32_phy32 gnu
```
but `bl_date.conf` is not updated. Why?

jkbk2004 commented:
> `test_changes.list' shows that these tests need new baselines:
> 
> ```
> control_gfs_mpas gnu
> rap_control intel
> rap_decomp intel
> rap_2threads intel
> rap_restart intel
> rap_sfcdiff intel
> rap_sfcdiff_decomp intel
> rap_sfcdiff_restart intel
> rap_control_dyn32_phy32 intel
> rap_2threads_dyn32_phy32 intel
> rap_restart_dyn32_phy32 intel
> rap_control gnu
> rap_decomp gnu
> rap_2threads gnu
> rap_restart gnu
> rap_sfcdiff gnu
> rap_sfcdiff_decomp gnu
> rap_sfcdiff_restart gnu
> rap_control_dyn32_phy32 gnu
> rap_2threads_dyn32_phy32 gnu
> rap_restart_dyn32_phy32 gnu
> ```
> 
> but `bl_date.conf` is not updated. Why?

For rap cases, we are cutting forecast time from 24 to 12. So, technically we just need to add some 9 and 12 hour files. We can further clean up 24 hour files when we update bl date in next round. @gspetro-NOAA any comment?

gspetro-NOAA commented:
@DusanJovic-NOAA It is as @jkbk2004 said--just adding new files to the baselines, not overwriting any of the current baseline files for the RAP tests.

DusanJovic-NOAA commented:
How are you 'just adding' new files? Manually?

DusanJovic-NOAA commented:
How do I know which files in which directories are new? Are new directories/files listed somewhere?

jkbk2004 commented:
Ok! we will go for new bl_date to avoid any confusion. @FernandoAndrade-NOAA @BrianCurtis-NOAA FYI

FernandoAndrade-NOAA commented:
I'm running into persistent out of memory errors with the `control_gfs_mpas_gnu` on Ursa during runtime.
please refer to:
`/scratch3/NAGAPE/epic/Fernando.Andrade-maldonado/reg-test/wm/ursa/ufs-weather-model/tests`
and `/scratch3/NAGAPE/epic/Fernando.Andrade-maldonado/stmp/ursa/RT_RUNDIRS/Fernando.Andrade-maldonado/FV3_RT/rt_3488999/control_gfs_mpas_gnu/err`

@jkbk2004 FYI

jkbk2004 commented:
@FernandoAndrade-NOAA we need to remove ursa option in rt.conf. Please, create a git issue. And we need to move on for merging.

FernandoAndrade-NOAA commented:
> @FernandoAndrade-NOAA we need to remove ursa option in rt.conf. Please, create a git issue. And we need to move on for merging.

Issue #2876 created.

jkbk2004 commented:
All tests are done ok. We can start merging process for this pr.

jkbk2004 commented:
@dustinswales new hash https://github.com/NOAA-EMC/fv3atm/commit/2f7f432ccbfabd6531941c08380fd2f235a098f0

jkbk2004 commented:
@dustinswales new CMEPS hash https://github.com/NOAA-EMC/CMEPS/commit/61b5a22ff810a647587c3e697b48a15c016f27d7

NickSzapiro-NOAA commented on driver/UFSDriver.F90:
```diff
@@ -23,13 +23,13 @@ MODULE UFSDriver
 !              /|\
 !             / | \
-!          ATM/OCN/ICE/WAV/LND/IPM/HYD/FIR .. components
+!          UFSATM/OCN/ICE/WAV/LND/IPM/HYD/FIR .. components
```
This should go back to ATM

NickSzapiro-NOAA replied:
This is minor and marking as resolved so can merge but hopefully can change in next round
